week4/weather.py

```diff
@@ -0,0 +1,123 @@
+import requests
+import matplotlib.pyplot as plt
+import random
+
+
+class AbstractDataService:
+    def get_data(self):
+        raise NotImplementedError("Subclasses must implement this method")
+
+class MockedDataService(AbstractDataService):
+    def get_data(self):
+        # Mocked data retrieval logic
+        mocked_data = {
+            'time': [],
+            'temperature_2m': []
+        }
+
+        start_temperature = 5.0
+        end_temperature = 20.0
+
+        # Generate mocked time and temperature data
+        for i in range(72):  # Assuming 72 data points based on the provided results
+            time = f'2023-11-26T{str(i).zfill(2)}:00'
+            temperature = round(random.uniform(start_temperature, end_temperature), 1)
+
+            mocked_data['time'].append(time)
+            mocked_data['temperature_2m'].append(temperature)
+
+        return mocked_data
+    
+    
+class APICallingService:
+    def call_api(self, payload: dict, url: str):
+        #return {"api_key": "api_value", "payload": payload}
+        r = requests.get(url, params=payload)
+        json_data = r.json()
+        hourly_data = json_data.get("hourly", {})
+        # Implement logic to call the external API with the provided payload
+        # Return the API response
+        #return {"api_key": "api_value"}
+        return hourly_data
+    
+
+class DataSourceHandler:
+    def __init__(self, data_service):
+        self.data_service = data_service
+
+    def handle_data(self):
+        data = self.data_service.get_data()
+        # Implement handling logic
+        print("Handling data:")
+        print(data)
+        return data
```

Comment:
similar to above, this method should expect parameters for the request (i.e. requested time range)